index.rst

```diff
@@ -41,40 +41,38 @@ with version control systems.
 File Format
 ===========
 
-EditorConfig files use an INI format that is compatible with the format used
-by `Python configparser Library`_. In an EditorConfig file (usually named
-``.editorconfig``), all beginning whitespace in each line is ignored. Each
-line must be one of the following:
-
-- Empty Line: An empty string.
-- Comment: A line starting with a ``;`` or a ``#``.
-- Section Title: A line that starts with a ``[``, and that ends with a ``]``.
-- Assignment: A line separated by an `=` into two parts.
-
-Any line that is not one of the above is invalid. "Inline comment" in
-appearance, i.e., a line that has a ``;`` or a ``#`` in the middle, is
-undefined. For convenience, we also define the following terminologies:
-
-- Section Name: The string between the beginning ``[`` and the ending ``]`` in
-  a Section Title.
-- Section: The lines starting from a Section Title until the beginning of the
-  next Section Title or end of file.
-- Property: The part before the first `=` in an Assignment.
-- Value: The part after the first `=` in an Assignment.
-
-In EditorConfig files, the Section Names are filepath globs, similar to the
-format accepted by gitignore. ``[`` and ``]`` are allowed in the Section
-Names. Forward slashes (``/``) are used as path separators and Backslashes
-(``\\``) are not allowed as path separators (even on Windows). Comments should
-always go on individual lines. EditorConfig files should be UTF-8 encoded, with
-either CRLF or LF line separators.
+EditorConfig uses an INI file format. In an EditorConfig file (usually named
+``.editorconfig``), all beginning whitespace on each line is considered
+irrelevant. Each line must be one of the following:
+
+- Blank: contains only whitespace characters.
+- Comment: starts with a ``;`` or a ``#``.
+   - Inline comments are not supported.
```

Comment:
I moved this point as a sub-point of the `Comment` bullet here so that the reader has more context. I continued this paradigm throughout the PR.

Reply:
I feel "supported" is a somewhat ambiguous word.  Plus the definition of "Inline comments" is gone.

Reply:
How about this?

> Inserting a `#` or `;` after non-whitespace characters in a line (i.e., inline) will not be parsed as a comment; in fact, it will be considered part of the section name or property in which you inserted it, which might not be what you intended.

Reply:
I think we have left this undefined as in https://github.com/editorconfig/editorconfig-vote/issues/6 . How about this?

> Inserting a `#` or `;` after non-whitespace characters in a line (i.e., inline) is undefined at this moment but reserved.

Reply:
What does undefined mean, exactly? How would that parse into AST, for example?

I'm left puzzled by [the remaining two tests](https://github.com/editorconfig/editorconfig-core-test/blob/48610d43b7455af12195473377f93c4ceea654f5/parser/comments.in#L47-L53):

```ini
# Octothorpe at end of value read as part of value
[test11.c]
key=value# not comment
```
☝️ notice how GitHub grays out the inline comment.

If `# not comment` is not a comment, then what is it? Undefined? Can we please define undefined in this context? Also, since we're trimming keys and values, would the value of `foo = bar # baz` be `bar` or `bar ` with a space before the comment indicator? We really need to 🔒 this down!

```ini
# Escaped octothorpe in value
[test12.c]
key=value \# not comment
```
☝️ notice again how GitHub grays this one out too, confusingly.

Reply:
I remember i've read parser-descriptions calling lines starting with a "comment"-character being ignored.

In that case the AST just wouldn't contain them as well as blank lines. Having ignored lines in the AST allows further use cases, like comment-linting or even more sophisiticated use cases (semantic use of comments like f.e. macros). This is probably not inteded for ecfg but for an INI-file in general it may be interesting?

Reply:
Unfortunately, there is no way to reproduce the original document without saving these values in the AST. Quite technically, what you are asking for is a concrete syntax tree (CST), which does not contain “irrelevant” content, including blank lines and comments.

Do you propose that we have both an AST as well as a CST representation of the document?

Reply:
Also, I’m not sure that being “ignored” means it doesn’t become part of the AST. It sounds to me like it means the implementation of that AST are just ignoring those values, which is consistent with this specification as well.

Reply:
Oh and sorry for the confusing reply: We shouldn't really do not recognize inline comments: Different Ini-Parsers may behave differently since in "real" Ini-files there might be values containing octothorpes?

Reply:
> Also, I’m not sure that being “ignored” means it doesn’t become part of the AST. It sounds to me like it means the implementation of that AST are just ignoring those values, which is consistent with this specification as well.

Good point - i thought we would anyways keep everything "we don't use" in the AST. So if somebody else want's to implement inline-comments he can do easily on top?

Reply:
Correct. Different INI passers definitely do different things. I don’t think we support inline comments, so we have to decide what to do with inline comment indicators and I’m still getting mixed or confusing feedback on this one item.

Reply:
The confusing part is that we’re saying we don’t support inline comments; yet, we’re saying in the same breath that if you do it it will be ignored? That sounds an awful lot like we’re supporting them, because ignoring what comes after a comment indicator is exactly the definition of a comment!

Reply:
Ahh - i see. I just implicitly thought, if the line is not introduced by a significant character (`# ; [`) it will be handled like a normal key-value line. Saying the inline-comments will become part of the value.

Since this might be a breaking change - we should

1. detect this in the transition-phase of editiorconfig-spec (and handle it accordingly after the AST-generation) and throw a warning **or**
2. allow that in the editorconfig-spec in general (and handle it accordingly after the AST-generation) **or**
3. allow it for editorconfig-ini and handle it in the AST-generation.

---
I hope you can handle that my statements iterate with my level of understanding..

Reply:
Oh and i 'd favor option 1 or 2

Reply:
The current EC tests, if I understand correctly, do not allow inline comments; thus, treat them as part of the key or value, depending on where you put them. That's different than being ignored though, isn't it?

Again, I don't think we need to introduce a breaking change to add clarity.

As such, I personally think we should be backwards compatible and clarify that:

> Inserting a `#` or `;` after non-whitespace characters in a line (i.e., inline) will not be parsed as a comment; in fact, it will be considered part of the section name or key-value pair in which it was inserted, which might not be the author's intent.

Comments should be handled in the formal grammar, which ends up in the AST.

If we do want to support inline comments, that would be a breaking change and we'd need to create an issue about that. For now, I just want to add clarity to the existing spec as it relates to the existing EC tests.

Reply:
It's difficult for me to argument the right way - i have the feeling i'm mixing up the EC-INI-spec with the EC-spec. Would it be heretic thinking about creating an EC-INI-parser creating an AST and using the EC-parser to transform into a EC-specific variant?

Reply:
Like having `[INI-file] -> [INI-AST] -> [EC-AST]`?

Reply:
Besides of that do i have the strong need to make a tabular comparison of different INI-parsers (which is probably not needed for the EC-spec on its own). And if we decide to go for an explicit EC-INI-parser, too, i'll try to collect some INI-stats as well.

Reply:
I don't think that makes sense, because EditorConfig is intended to be flexible in that it supports any custom declarations you throw at it. Let me try to help you draw the line between the AST and EC's logic though:

## INI Parsing into AST

The INI parser just handles the parsing of the INI file, strictly according to the INI specification. In theory, it wouldn't need to know anything about EditorConfig, because it doesn't have any business logic in it to that effect. This means the INI AST can be used by anyone in contexts that have nothing to do with EditorConfig.

## EditorConfig Business Logic

- Merging INI files until `root=true` is found (the final merge might be the EC AST you were talking about).
- Parsing the section headers as filepath globs.
- Ignoring declarations that are not directly supported by EditorConfig, even though they exist in the AST.
- In the future, throwing validation errors that are specific to EditorConfig and not to the INI parsing (e.g., `indent_size` cannot equal `two` and expected an integer value).

Reply:
I see - thanks for the clarification! It's hard for me to create an opinion about something as controversial¹ as inline-comments.

¹) I snooped into some INI-parser-implementations and many of them treat the different or provide comment-specific modes. It seems to be the same with value-parsing, though. There seem to exist different parsers allowing to encapsulate the value in double-quotes allowing to have significants in the value - others seem to have not.

For me this is at first a INI-grammar-specific problem.

Reply:
According to editorconfig/editorconfig-vote#6 , I think we have left the decision of this task to a future date. We neither treat text after a semicolon or sharp as part of the value nor comment. The rationale is that we haven't encountered a situation in which we should make a decision with respect to this issue. It likely leaves us more flexibility and leeway when we need to make a decision because there is a need. This is also pretty common strategy in the development of specs in other areas. For now, we can just say, users should not use the so-called "inline comments". We do not guarantee what will happen at this point.

Reply:
> I'm left puzzled by [the remaining two tests](https://github.com/editorconfig/editorconfig-core-test/blob/48610d43b7455af12195473377f93c4ceea654f5/parser/comments.in#L47-L53):
> 
> ```ini
> # Octothorpe at end of value read as part of value
> [test11.c]
> key=value# not comment
> ```
> 
> point_up notice how GitHub grays out the inline comment.
> 
> If `# not comment` is not a comment, then what is it? Undefined? Can we please define undefined in this context? Also, since we're trimming keys and values, would the value of `foo = bar # baz` be `bar` or `bar ` with a space before the comment indicator? We really need to lock this down!
> 
> ```ini
> # Escaped octothorpe in value
> [test12.c]
> key=value \# not comment
> ```
> 
> point_up notice again how GitHub grays this one out too, confusingly.

These two tests are probably something we forgot to remove after https://github.com/editorconfig/editorconfig-vote/issues/6 . They are testing what https://github.com/editorconfig/editorconfig-vote/issues/6 exactly opposes to.

I don't think GitHub grays it out means anything (plus its buggy syntax highlighting). People define INI differently. For example, [configparser](https://docs.python.org/3/library/configparser.html) says:

> Configuration files may include comments, prefixed by specific characters (# and ; by default 1). **Comments may appear on their own on an otherwise empty line, possibly indented.**

"Undefined" might be better named "implementation-defined": "No warranty, users should avoid using it. How a so-called 'inline comment' is handled is not uniform across implementation at this moment."  As commented above, this is to leave more flexibility when we truly need to make a decision about this issue for another new feature in the future.

Reply:
Resolved in https://github.com/editorconfig/specification/pull/12/commits/3d96e629245a53844b1cd64611cfdf054265c9e7.